File: Python/HexBinaryCalc.py

HEX_TO_BIN: dict = {
    "0":"0000", "1":"0001", "2":"0010", "3":"0011",
    "4":"0100", "5":"0101", "6":"0110", "7":"0111",
    "8":"1000", "9":"1100", "A":"1001", "B":"1010",
    "C":"1011", "D":"1100", "E":"1110", "F":"1111"
}


def convertHexToBinary(hexInput: str) -> str:

    bin: str = ''

    for s in hexInput[2:]:
        bin += HEX_TO_BIN[s.upper()]
    return bin 

def convertBinaryToHex(BinaryInput: str) -> str:
    """

        --> Check if bin number input by user is divisible by 4. if not, add the remainder of numbers needed to make it divisible by 4 to the 
        left of the binary number!

        --> Split the bin number into chunks of 4, and then convert it to hex values w/ base 2. 
    """
    hex_output: str = ''

    for index in range(0, len(BinaryInput), 4):
        chunk = BinaryInput[index:index + 4]

        if len(chunk) != 4:
            chunk = chunk.zfill(4)

        for hex_digit, binary_value in HEX_TO_BIN.items():
            if binary_value == chunk:
                hex_output += hex_digit
                break

    return "0x" + hex_output


def isvalidateHex(string: str) -> bool:

    #Valid hex: 0 - 9, a - f or A - F

    if len(string) <= 2:
        return False 
    if not string.startswith(('0x', '0X')):
        return False
    return all(s in '0123456789abcdefABCDEF' for s in string[2:])

def isvalidBinary(string: str) -> bool: 
    
    if string == "":
        return False
    
    return all(c in ('0', '1') for c in string)

def getConversionDirection() -> str:
    
    print("======= Hex/Binary Converter =======")
    choice: str = input("Please input desired conversion (hex/binary):\n\n")

    choice = choice.lower() # Note: Strings in python are IMMUTABLE (can not be modified in place aka. simply doing choice.lower() is incorrect)

    while choice not in ('hex', 'binary'):
        choice = input("Invalid conversion type, please try again:\n\n")
        choice = choice.lower()

    return choice

def main():
    
    try:
        
        direction: str = getConversionDirection()
        result: str = ""
        userInput: str = ""

        if direction == 'hex':
            userInput = input("Enter a Binary value:    ")

            while not isvalidBinary(userInput):
                userInput = input("Invalid binary input, please try again:\n\n")
            
            result = convertBinaryToHex(userInput)
            print(f"Binary value: {userInput} \n Hex value: {result}")

        else:

            userInput = input("Enter a Hex value:   ")

            while not isvalidateHex(userInput):
                userInput = input("Invalid hex input, please try again:\n\n")

            result = convertHexToBinary(userInput)
            print(f"Hex value: {userInput} \n Binary value: {result}")

    except (ValueError, TypeError) as e:
        print("Error: ", e)


if __name__ == "__main__":
    main()

File: Cpp/HexBinaryCalc.cpp
#include <iostream>
#include <map>
#include <string>
#include <algorithm>
#include <stdexcept>
#include <cctype>

// Conversion maps
const std::map<char, std::string> HEX_TO_BIN = {
    {'0', "0000"}, {'1', "0001"}, {'2', "0010"}, {'3', "0011"},
    {'4', "0100"}, {'5', "0101"}, {'6', "0110"}, {'7', "0111"},
    {'8', "1000"}, {'9', "1001"}, {'A', "1010"}, {'B', "1011"},
    {'C', "1100"}, {'D', "1101"}, {'E', "1110"}, {'F', "1111"}
};

//converts hexadecimal string to binary representation
std::string convertHexToBinary(const std::string& hexInput){

    std::string result = ""; 
    /* Note, you could also do this: 

        std::string hexDigits = hexInput.substr(2); (to remove 0x)
        for(char c : hexDigits){

            result += HEX_TO_BIN.at(c)
        }


    */
    for(size_t i = 2; i < hexInput.size(); ++i){
        result += HEX_TO_BIN.at(hexInput[i]);
    }

    return result; 
}

// Converts binary string to hexadecimal representation
std::string convertBinaryToHex(const std::string& binaryInput) {
    
    std::string result = ""; 

    std::string bin = binaryInput;
    int raw = static_cast<int>(bin.size() % 4); 
    //if bin.size %4 != 0, add 2 zeros on left most side
    if(raw != 0)  bin = std::string(4 - raw, '0') + bin; // Or equivalently; bin.insert(0, 4- raw, '0'); 

    for(size_t i = 0; i < bin.size(); i+=4){
        std::string chunk = bin.substr(i, 4);
        int val = std::stoi(chunk, nullptr, 2);
        char hexChar; 
        if (val < 10) hexChar = static_cast<char>('0' + val);
        else hexChar = static_cast<char>('A' + (val - 10));
        result += hexChar; 
    }

    return std::string("0x") + result;
}


// Validates if a string is a valid hexadecimal number (with 0x prefix)
bool isValidHexadecimal(const std::string& input) {
    if (input.length() <= 2) return false;
    if ((input.substr(0, 2) != "0x" && input.substr(0, 2) != "0X")) return false;   // substr() definitiely a worthy function to keep in mind. 
    
    return std::all_of(input.begin() + 2, input.end(),  // all_of is definitely another worthy funtion. 
        [](char c) { return (c >= '0' && c <= '9') || 
                            (c >= 'a' && c <= 'f') || 
                            (c >= 'A' && c <= 'F'); });
}

// Validates if a string contains only binary digits (0 or 1)
bool isValidBinary(const std::string& input) {
    if (input.empty()) return false;
    
    return std::all_of(input.begin(), input.end(),
        [](char c) { return c == '0' || c == '1'; });
}

// Prompts user to select conversion direction
std::string getConversionDirection() {
    std::string choice;
    
    std::cout << "\n=== Hex/Binary Converter ===\n";
    std::cout << "Convert to (hex/binary): ";
    std::cin >> choice;
    
    // Convert to lowercase for case-insensitive comparison
    std::transform(choice.begin(), choice.end(), choice.begin(), ::tolower);
    
    while (choice != "hex" && choice != "binary") {
        std::cout << "Invalid input. Please enter 'hex' or 'binary': ";
        std::cin >> choice;
        std::transform(choice.begin(), choice.end(), choice.begin(), ::tolower);
    }
    
    return choice;
}

int main() {

    try {
        std::string direction = getConversionDirection();
        std::string input;
        std::string result;
        
        if (direction == "hex") {
            // Convert binary to hex
            std::cout << "\nEnter binary value: ";
            std::cin >> input;
            
            while (!isValidBinary(input)) {
                std::cout << "Invalid binary input. Please enter only 0s and 1s: ";
                std::cin >> input;
            }
            
            result = convertBinaryToHex(input);
            std::cout << "\nBinary: " << input << "\nHexadecimal: " << result << "\n";
            
        } else {
            // Convert hex to binary
            std::cout << "\nEnter hexadecimal value (e.g., 0xABCD): ";
            std::cin >> input;
            
            while (!isValidHexadecimal(input)) {
                std::cout << "Invalid hexadecimal input. Please enter with 0x prefix: ";
                std::cin >> input;
            }
            
            result = convertHexToBinary(input);
            std::cout << "\nHexadecimal: " << input << "\nBinary: " << result << "\n";
        }
        
    } catch (const std::exception& e) {
        std::cerr << "Error: " << e.what() << "\n";
        return 1; // Note: returning 1 in main signifies that the program has come across an error! But in other functions, returning 1 = true. 
    }
    
    return 0;
}